### DISCUSSION TITLE
Cannot find headers for set prefix

### DISCUSSION THREAD

**Discussion by albinahlback** [ImageMagick/ImageMagick#5782] [Help]
### ImageMagick version

Nightly

### Operating system

Linux

### Operating system, version and so on

Arch Linux

### Description

After installing ImageMagick from source, I supposed that I try out some [MagickWand examples](https://imagemagick.org/script/magick-wand.php). However, GCC directly complains that
```
/my/prefix/ImageMagick-7/MagickWand/MagickWand.h:29:12: fatal error: MagickCore/magick-config.h: No such file or directory
   29 | #  include "MagickCore/magick-config.h"
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

It is obvious that GCC can't find `MagickCore/magick-config.h` as it is not under `/my/prefix`, but rather `/my/prefix/ImageMagick-7`.

Is this expected? What input to `configure` should I give so that I do not have to link the directory `/my/prefix/ImageMagick-7` everytime?

### Steps to Reproduce

```
$ cd /path/to/ImageMagick_source
$ ./configure --prefix=/my/prefix --enable-opencl --enable-static=no --disable-deprecated --with-fftw CC=gcc
$ make
$ make install
$ echo "#include <ImageMagick-7/MagickWand/MagickWand.h>" >> testfile.c
$ echo "int main(){return 0;}" >> testfile.c
$ gcc testfile.c
```

**Comment by snibgo** [comment]
> $ cd /path/to/ImageMagick_source

What do you have in that directory? I expect you have `Makefile`, and directories `MagickCore`, `MagickWand`, `filters`, and so on, with no *.c or *.cpp. If you don't, then cd to the directory with those files. Then `./configure`, `make` and `make install` should work. I call that directory `IM7SRC`.

When I write standalone programs for IM, I put them in a subdirectory of that directory, named `${IM7SRC}\snibgo`. Where did you put your source code? Perhaps you didn't use a subdirectory.

Programs include lines like:
```
#include <MagickWand/MagickWand.h>
```
To build them, I use three bash commands from ${IM7SRC}:
```
export PKG_CONFIG_PATH=${IM7DEV}../lib/pkgconfig

cc -o "$1" "$1".c -Wall -Wextra -Wunused-variable -Warray-bounds -Wconversion -Wanalyzer-use-after-free -Wpedantic -Wunused-parameter -Warray-parameter -Wdeprecated -march=corei7 -mtune=corei7 -Wa,-march=corei7,-mtune=corei7 `pkg-config --cflags --libs MagickWand`

cp "$1".exe ${IM7DEV}
```
... where $1 is `snibgo\sigw` or whatever.

The `pkg-config` varies if I need just MagickCore.

**Comment by albinahlback** [comment]
> > $ cd /path/to/ImageMagick_source
> 
> What do you have in that directory?

Everything contained in the ImageMagick repository.

> If you don't, then cd to the directory with those files. Then `./configure`, `make` and `make install` should work. I call that directory `IM7SRC`.

Everything I did followed from `Install-unix.txt`. Do I really need to rename the ImageMagick source directory?

> When I write standalone programs for IM, I put them in a subdirectory of that directory, named `${IM7SRC}\snibgo`. Where did you put your source code? Perhaps you didn't use a subdirectory.

Why would I need to do this? Perhaps I'm not familiar enough with UNIX systems, but I'm pretty sure that the convention is that `./configure --prefix=my/prefix && make && make install` will first configure, then compile and then install under `my/prefix`. Then I should be able to include any installed header such as `MagickWand/MagickWand.h`.

> Programs include lines like:
> 
> ```
> #include <MagickWand/MagickWand.h>
> ```
> 
> To build them, I use three bash commands from ${IM7SRC}:
> 
> ```
> export PKG_CONFIG_PATH=${IM7DEV}../lib/pkgconfig
> 
> cc -o "$1" "$1".c -Wall -Wextra -Wunused-variable -Warray-bounds -Wconversion -Wanalyzer-use-after-free -Wpedantic -Wunused-parameter -Warray-parameter -Wdeprecated -march=corei7 -mtune=corei7 -Wa,-march=corei7,-mtune=corei7 `pkg-config --cflags --libs MagickWand`
> 
> cp "$1".exe ${IM7DEV}
> ```
> 
> ... where $1 is `snibgo\sigw` or whatever.
> 
> The `pkg-config` varies if I need just MagickCore.

Why do I need to bother with `pkg-config`? It is not mentioned under the installation notes and my understanding is that it is not a convention to use it.

**Comment by snibgo** [comment]
> > > $ cd /path/to/ImageMagick_source

> > What do you have in that directory?

> Everything contained in the ImageMagick repository.

That doesn't answer my question. Does it contain `Makefile`, and directories `MagickCore`, `MagickWand`, `filters`, and so on?

> Do I really need to rename the ImageMagick source directory?

You don't need to rename any directories. I don't. I use environment variable `${IM7SRC}` to refer to the directory that contains Makefile and the directories I mentioned. I need to build and install multiple versions of IM, and I don't want my scripts to include absolute directory paths, so I use environment variables: `${IM7SRC}` is the directory that contains Makefile, and directories MagickCore and MagickWand, etc.

> Perhaps I'm not familiar enough with UNIX systems, but I'm pretty sure that the convention is that ./configure --prefix=my/prefix && make && make install will first configure, then compile and then install under my/prefix.

Yes. That will install the IM binary in directory `my/prefix/bin`. When you build a standalone program, you should probably move the built binary to that directory.

> Why do I need to bother with pkg-config? It is not mentioned under the installation notes and my understanding is that it is not a convention to use it.

It is needed for building standalone programs. It is mentioned near the top of the page you linked to, https://imagemagick.org/script/magick-wand.php .

If I don't `export PKG_CONFIG_PATH=...` then I get an error:
```
snibgo\sigw.c:4:10: fatal error: MagickWand/MagickWand.h: No such file or directory
    4 | #include <MagickWand/MagickWand.h>
      |          ^~~~~~~
compilation terminated.
```
You get a similar error.

**Comment by albinahlback** [comment]
> That doesn't answer my question. Does it contain `Makefile`, and directories `MagickCore`, `MagickWand`, `filters`, and so on?

Ah, sorry. Yes, it does. The whole installation process seems to work fine, I checked the configure log and everything seemed to be okay.

> You don't need to rename any directories. I don't. I use environment variable `${IM7SRC}` to refer to the directory that contains Makefile and the directories I mentioned. I need to build and install multiple versions of IM, and I don't want my scripts to include absolute directory paths, so I use environment variables: `${IM7SRC}` is the directory that contains Makefile, and directories MagickCore and MagickWand, etc.

I don't have a need for multiple versions. Is there someway to omit the include path suffix `/ImageMagick-${MAJOR_VERSION}` without any hacking? I simply want it to copy the headers into `/my/prefix/include` and not `my/prefix/include/ImageMagick-${MAJOR_VERSION}`.

**Comment by snibgo** [comment]
> Is there someway to omit the include path suffix /ImageMagick-${MAJOR_VERSION} without any hacking? I simply want it to copy the headers into /my/prefix/include and not my/prefix/include/ImageMagick-${MAJOR_VERSION}.

I don't know how to do that. You could, of course, simply copy the files after building IM, if you really want to.

I don't know why you want to do that. `pkg-config` sets a `cc -I` option so the .h files in ImageMagick-7 will be found, without having to specify that subdirectory in `#include` statements.

**Comment by albinahlback** [comment]
I'm only after the C API, for building a library and I don't want to use some custom way to include headers.
